Include `elideBlocks` state in `Instantiate` hash

Currently, if you use `elideBlocks` on `Instantiate`, this will have no effect if there is a cache hit. Instead, whether or not you are in an `elideBlocks` scope should factor into the hash so that the following circuit will compile.

In the example below, this will error in `firtool` compilation because of bind-under-bind checking:

``` scala
//> using repository https://central.sonatype.com/repository/maven-snapshots
//> using scala 2.13.18
//> using dep org.chipsalliance::chisel:7.6.0
//> using plugin org.chipsalliance:::chisel-plugin:7.6.0
//> using options -unchecked -deprecation -language:reflectiveCalls -feature -Xcheckinit
//> using options -Xfatal-warnings -Ywarn-dead-code -Ywarn-unused -Ymacro-annotations

import chisel3._
import chisel3.experimental.hierarchy.{instantiable, public, Instantiate}
import circt.stage.ChiselStage

@instantiable
class Bar extends Module {
  @public
  val in = dontTouch(IO(Input(Bool())))
  layer.block(layers.Verification.Assert) {
    val x = dontTouch(Wire(Bool()))
    x :<= DontCare
  }
}

class Foo extends Module {
  val bar = Instantiate(new Bar)
  layer.block(layers.Verification) {
    val bar = layer.elideBlocks {
      Instantiate(new Bar)
    }
    bar.in :<= dontTouch(WireInit(true.B))
  }
}

object Main extends App {
  println(
    ChiselStage.emitSystemVerilog(
      gen = new Foo,
      firtoolOpts = Array(
        "-disable-layers=Verification.Assume,Verification.Cover",
        "-strip-debug-info"
      )
    )
  )
}
```

Error:

```
firtool returned a non-zero exit code. Note that this version of Chisel (7.6.0) was published against firtool version 1.138.0.
------------------------------------------------------------------------------
ExitCode:
1
STDOUT:

STDERR:
/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/Foo.scala:12:2: error: 'firrtl.module' op either contains layerblocks or has at least one instance that is or contains a Grand Central companion or layerblocks
@instantiable
 ^
/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/Foo.scala:12:2: note: see current operation: 
"firrtl.module"() <{annotations = [], convention = #firrtl<convention internal>, domainInfo = [], layers = [], portAnnotations = [], portDirections = array<i1: false, false, false>, portLocations = [loc("/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/ankit.scala":12:2), loc("/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/ankit.scala":12:2), loc("/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/ankit.scala":15:24)], portNames = ["clock", "reset", "in"], portSymbols = [], portTypes = [!firrtl.clock, !firrtl.reset, !firrtl.uint<1>], sym_name = "Bar"}> ({
^bb0(%arg0: !firrtl.clock, %arg1: !firrtl.reset, %arg2: !firrtl.uint<1>):
  "firrtl.skip"() : () -> ()
  "firrtl.layerblock"() <{layerName = @Verification}> ({
    "firrtl.layerblock"() <{layerName = @Verification::@Assert}> ({
      %0 = "firrtl.wire"() <{annotations = [], name = "x", nameKind = #firrtl<name_kind interesting_name>}> : () -> !firrtl.uint<1>
      %1 = "firrtl.invalidvalue"() : () -> !firrtl.uint<1>
      "firrtl.matchingconnect"(%0, %1) : (!firrtl.uint<1>, !firrtl.uint<1>) -> ()
    }) : () -> ()
  }) : () -> ()
}) {sym_visibility = "private"} : () -> ()
/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/Foo.scala:16:43: note: illegal layerblock here
  layer.block(layers.Verification.Assert) {
                                          ^
/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/Foo.scala:16:43: note: illegal layerblock here
/Users/schuylereldridge/repos/github.com/seldridge/scala-snippets/Foo.scala:26:18: note: illegal instantiation under a layerblock here
      Instantiate(new Bar)
```

--- a/core/src/main/scala/chisel3/experimental/hierarchy/core/Definition.scala
+++ b/core/src/main/scala/chisel3/experimental/hierarchy/core/Definition.scala
@@ -121,7 +121,8 @@ object Definition extends SourceInfoDoc {
         context.contextCache,
         context.layerMap,
         context.inlineTestIncluder,
-        context.suppressSourceInfo
+        context.suppressSourceInfo,
+        context.elideLayerBlocks
       )
     }
     dynamicContext.inDefinition = true

--- a/core/src/main/scala/chisel3/internal/Builder.scala
+++ b/core/src/main/scala/chisel3/internal/Builder.scala
@@ -477,7 +477,8 @@ private[chisel3] class DynamicContext(
   val contextCache:       BuilderContextCache,
   val layerMap:           Map[layer.Layer, layer.Layer],
   val inlineTestIncluder: InlineTestIncluder,
-  val suppressSourceInfo: Boolean
+  val suppressSourceInfo: Boolean,
+  var elideLayerBlocks:   Boolean
 ) {
   val importedDefinitionAnnos = annotationSeq.collect { case a: ImportDefinitionAnnotation[_] => a }
 
@@ -545,12 +546,11 @@ private[chisel3] class DynamicContext(
   var blockStack:           List[Block] = Nil
   // Clock and Reset are "Delayed" because ImplicitClock and ImplicitReset need to set these values,
   // But the clock or reset defined by the user won't yet be initialized
-  var currentClock:     Option[Delayed[Clock]] = None
-  var currentReset:     Option[Delayed[Reset]] = None
-  var currentDisable:   Disable.Type = Disable.BeforeReset
-  var enabledLayers:    mutable.LinkedHashSet[layer.Layer] = mutable.LinkedHashSet.empty
-  var layerStack:       List[layer.Layer] = layer.Layer.root :: Nil
-  var elideLayerBlocks: Boolean = false
+  var currentClock:   Option[Delayed[Clock]] = None
+  var currentReset:   Option[Delayed[Reset]] = None
+  var currentDisable: Disable.Type = Disable.BeforeReset
+  var enabledLayers:  mutable.LinkedHashSet[layer.Layer] = mutable.LinkedHashSet.empty
+  var layerStack:     List[layer.Layer] = layer.Layer.root :: Nil
   val errors = new ErrorLog(warningFilters, sourceRoots, throwOnFirstError)
   val namingStack = new NamingStack
 

--- a/src/main/scala/chisel3/stage/phases/Elaborate.scala
+++ b/src/main/scala/chisel3/stage/phases/Elaborate.scala
@@ -55,7 +55,8 @@ class Elaborate extends Phase {
             BuilderContextCache.empty,
             chiselOptions.layerMap,
             chiselOptions.inlineTestIncluder,
-            chiselOptions.suppressSourceInfo
+            chiselOptions.suppressSourceInfo,
+            false
           )
         val (elaboratedCircuit, dut) = {
           Builder.build(Module(gen()), context)

--- a/src/test/scala-2/chiselTests/experimental/hierarchy/DefinitionSpec.scala
+++ b/src/test/scala-2/chiselTests/experimental/hierarchy/DefinitionSpec.scala
@@ -924,4 +924,25 @@ class DefinitionSpec extends AnyFunSpec with Matchers with FileCheck {
         )
     }
   }
+  describe("(8): A Definition under an elideBlock scope") {
+    it("(8.a): should elide layer blocks") {
+      class Bar extends RawModule {
+        layer.block(layers.Verification) {
+          val a = WireInit(false.B)
+        }
+      }
+      class Foo extends RawModule {
+        layer.elideBlocks {
+          val barDef = Definition(new Bar)
+        }
+      }
+      ChiselStage
+        .emitCHIRRTL(new Foo)
+        .fileCheck()(
+          """|CHECK: module Bar :
+             |CHECK-NOT: layerblock
+             |""".stripMargin
+        )
+    }
+  }
 }

--- a/src/test/scala-2/chiselTests/experimental/hierarchy/InstantiateSpec.scala
+++ b/src/test/scala-2/chiselTests/experimental/hierarchy/InstantiateSpec.scala
@@ -535,8 +535,14 @@ class InstantiateSpec extends AnyFunSpec with Matchers with FileCheck {
       }
 
       emitCHIRRTL(new Foo).fileCheck()(
-        """|CHECK-COUNT-3: module
-           |CHECK-NOT:     module
+        """|CHECK:     module HasLayer :
+           |CHECK-NOT: module
+           |CHECK:       layerblock
+           |
+           |CHECK:     module HasLayer{{.*}} :
+           |CHECK-NOT:   layerblock
+           |
+           |CHECK:     module Foo :
            |""".stripMargin
       )
     }
@@ -551,9 +557,12 @@ class InstantiateSpec extends AnyFunSpec with Matchers with FileCheck {
         }
       }
 
-      emitCHIRRTL(new Foo).fileCheck()(
-        """|CHECK-COUNT-2: module
-           |CHECK-NOT:     module
+      emitCHIRRTL(new Foo).fileCheck("--implicit-check-not=module")(
+        """|CHECK:     module HasLayer :
+           |CHECK-NOT:   layerblock
+           |
+           |CHECK:     module Foo :
+           |CHECK-NOT:   layerblock
            |""".stripMargin
       )
     }

